### PR TITLE
Native Backend Getting Start

### PR DESCRIPTION
**Pull request by jokerwyt**
```
python compiler/element_compiler_test.py --element examples/elements/echo_elements/empty.appnet --backend envoy_native --placement client --proto ping.proto --method_name PingEcho --verbose

# to compile and run the generated Envoy
/mnt/compiler/compiler/element/generated/envoy_native$ ./run.sh 
```

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[submodule "compiler/element/backend/envoy_native/template"]
+	path = compiler/element/backend/envoy_native/template
+	url = git@github.com:jokerwyt/envoy-appnet.git
```
Comment (Romero027) [inline .gitmodules:3]: can you move this to appnet org?

File: compiler/element/backend/envoy_native/appnettype.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Created and used by Envoy natvie backend.
+
+class AppNetType:
```
Comment (Romero027) [inline compiler/element/backend/envoy_native/appnettype.py:3]: consider rename these to NativeTypes (incl. filename)

File: compiler/element/backend/envoy_native/cppgen.py
Comment (Romero027) [inline compiler/element/backend/envoy_native/cppgen.py:1]: this file should be called nativegen.py

File: compiler/element/backend/envoy_native/appnettype.py (hunk 2 above)
Comment (Romero027) [reply compiler/element/backend/envoy_native/appnettype.py:3]: I also don't quite understand what's this is used for

**Review by Romero027 — CHANGES_REQUESTED** [review]
LGTM overall.

Can you remove all the code that you don't need? For example, there are many wasm code in the graph compiler directory.

Also, we need to keep the name consistent. Use Native instead of cpp for now.

File: compiler/element/backend/envoy_native/appnettype.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+
+from typing import Optional
+from compiler.element.logger import ELEMENT_LOG as LOG
+
+from compiler.element.backend.envoy_native.types import *
+
+
+
+# Every expression, state and temporary variables in AppNet belong to a AppNetType.
```
Comment (Kristoff-starling) [inline compiler/element/backend/envoy_native/appnettype.py:11]: Do you mean that this file is used for type inference? In this case, I suggest that it should be moved to the frontend/ir folder as it's backend-independent.

File: compiler/element/backend/envoy_native/appnettype.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
+# Created and used by Envoy natvie backend.
```
Comment (Kristoff-starling) [inline compiler/element/backend/envoy_native/appnettype.py:1]: Typo natvie -> native

File: compiler/element/backend/envoy_native/nativetype.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+from enum import Enum
+from typing import List, Optional
+
+from compiler.element.backend.envoy_native import *
+from compiler.element.logger import ELEMENT_LOG as LOG
+
+class NativeType:
+  def gen_decl(self, name: str) -> str:
+    LOG.error(f"gen_decl not implemented for {self}")
+    assert(0)
```
Comment (Kristoff-starling) [inline compiler/element/backend/envoy_native/nativetype.py:13]: I suggest using NotImplementedError as assert(0) looks weird :)

**Review by Kristoff-starling — CHANGES_REQUESTED** [review]
LGTM overall

@Romero027 is pre-commit manually turned off? I think it's helpful for keeping the code style consistent, e.g., indentation length

File: compiler/element/backend/envoy_native/types.py
Comment (Kristoff-starling) [inline compiler/element/backend/envoy_native/types.py:1]: This is a file that only imports classes?

**Comment by Romero027** [comment]
LGTM, feel free to merge after you address Yuyao's comment

**Review by Romero027 — APPROVED** [review]

File: compiler/element/backend/envoy_native/appnettype.py (hunk 3 above)
Comment (jokerwyt) [reply compiler/element/backend/envoy_native/appnettype.py:11]: Yes. But changing fronend/ir may cause problems in other backends, so I keep it in the native backend for now. It should be moved outside in future.

File: compiler/element/backend/envoy_native/types.py
Comment (jokerwyt) [reply compiler/element/backend/envoy_native/types.py:1]: Yes.

**Review by Kristoff-starling — APPROVED** [review]

**Pull request merged**